File: querydsl-libraries/querydsl-jpa/src/test/java/com/querydsl/jpa/domain/Money.java
package com.querydsl.jpa.domain;

import java.io.Serial;
import java.math.BigDecimal;
import java.util.Objects;

public class Money extends Number implements Comparable<Money> {

  @Serial private static final long serialVersionUID = 1L;

  private final BigDecimal value;

  public Money(BigDecimal value) {
    this.value = value;
  }

  public Money(String value) {
    this.value = new BigDecimal(value);
  }

  public BigDecimal getValue() {
    return value;
  }

  @Override
  public String toString() {
    return value.toPlainString();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Money)) return false;
    Money that = (Money) o;
    return Objects.equals(value, that.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public long longValue() {
    return value.longValue();
  }

  @Override
  public int intValue() {
    return value.intValue();
  }

  @Override
  public float floatValue() {
    return value.floatValue();
  }

  @Override
  public double doubleValue() {
    return value.doubleValue();
  }

  @Override
  public int compareTo(Money o) {
    return this.value.compareTo(o.value);
  }
}

File: querydsl-examples/querydsl-example-ksp-codegen/src/main/kotlin/com/querydsl/example/ksp/Money.kt
package com.querydsl.example.ksp

import java.math.BigDecimal

data class Money(val value: BigDecimal) : Number(), Comparable<Money> {

    override fun toByte(): Byte = value.toByte()
    override fun toDouble(): Double = value.toDouble()
    override fun toFloat(): Float = value.toFloat()
    override fun toInt(): Int = value.toInt()
    override fun toLong(): Long = value.toLong()
    override fun toShort(): Short = value.toShort()

    override fun compareTo(other: Money): Int = this.value.compareTo(other.value)
}

